### DISCUSSION TITLE
Copilot Enterprise Q & A

### DISCUSSION THREAD

**Discussion by 2percentsilk** [community/community#77982] [Enterprise]
Hey folks!

If you've got **questions about the new Copilot Enterprise offering**, please let us know and we'll do our best to assist 😄 

### Best Practices
- Before posting a question, please review the other questions/answers in this post to avoid duplication
- Be sure to review our [documentation](https://docs.github.com/en/copilot/github-copilot-enterprise/overview/about-github-copilot-enterprise) and [announcement post](https://github.blog/2023-11-08-universe-2023-copilot-transforms-github-into-the-ai-powered-developer-platform/#introducing-github-copilot-enterprise-copilot-personalized-to-your-organization).

### Copilot Enterprise - Private Beta
- **This is a private beta only open to GHEC Customers.** 
- To access the private beta, please [sign up for the waitlist](https://github.com/github-copilot/copilot_enterprise_waitlist_signup/join) 
- Interested enterprise customers can reach out to your GitHub account manager or [contact our sales team](https://github.com/enterprise/contact) for enablement.

#### This thread is just for Questions & Answers. 

We are **collecting feedback on the private beta** in another discussion: [Copilot Enterprise - Private Beta Feedback](https://github.com/orgs/community/discussions/77976)

Happy coding!

**Comment by Beanmasters** [comment]
A key driving feature in Copilot Enterprise is its ability to know the code base, not just what is open in the editor.  My understanding is that Copilot Enterprise currently requires a GitHub Enterprise Cloud repository for code base integration.  Are there plans to allow code base integration from other Git providers such as Azure DevOps Git repositories?

**Reply by Beanmasters** [reply]
Hello?

**Reply by arfon** [reply]
Thanks for the question! There are no immediate plans to integrate with other services, but we will definitely keep this feature in mind. Thanks again.

**Reply by Beanmasters** [reply]
Thanks for providing that update.  I attended the Microsoft "Accelerate Development using Azure DevOps and GitHub" webinar yesterday where they had a few slides and demos on GitHub CoPilot and asked:

_Is there an option to use GitHub Copilot Enterprise with an Azure DevOp repository on a local machine instead of Codespaces?_

Their response was:

_"Yes, you can use GitHub Copilot Enterprise with an Azure DevOps repository on a local machine by installing the Copilot extension in your IDE, such as Visual Studio Code. This allows Copilot to work with code in your local Azure DevOps Git repository."_

This seems to be counter to your answer, can you confirm what would be the case here?

**Reply by arfon** [reply]
My apologies, I thought you were asking if you would be able to use GitHub Copilot Enterprise (inside of github.com) for repositories outside of the GitHub.com hosted environment.

As our colleagues at Microsoft have pointed out, it's perfectly possible to use GitHub Copilot in the IDE with repositories from any provider. Hopefully that clears things up any further questions just let me know.

**Reply by Beanmasters** [reply]
OK, the ability to use CoPilot on a local development workstation customized with an Azure DevOps code repository (i.e. CoPilot knows about the entire code repository, not just what is open in the editor) is where my question is based.  The caveat with that setup, according to Microsoft, is:

_"GitHub Copilot Enterprise requires a GitHub Enterprise Cloud subscription, even if you are using Azure DevOps repositories. This requirement is in place because Copilot Enterprise features and licensing are tied to the GitHub Enterprise Cloud services."_

So cost-wise in this setup, we need to pay for two GitHub services but would only be utilizing CoPilot, which does not seem ideal.  GitHub Enterprise Cloud does offer other features, which, unfortunately, we already have similar services in place.

There is a CoPilot fine-tuning BETA for CoPilot Business that purports to "fine-tune a private, custom copilot using your own repositories", so seems that may be a better option than the above scenarios above.

Thanks for the additional information.

**Reply by schoesa** [reply]
Are there any news about this topic (integrating different code bases)?

**Comment by GytisZ** [comment]
Hi, this is a long awaited feature, very happy with the release.

We've upgraded to Copilot Enterprise. Copilot now helpfully comments on our indexed codebases. However, it's unclear whether this functionality is also available via IDEs?

I can't find anything in the update notes for the Copilot Chat or Copilot extensions and nothing mentioning Enterprise in the pre-release versions.

While web interface is nice, for us, the real value comes if we could use it in everyday workflows via IDE rather than switching back and forth between website and the IDE.

**Reply by timrogers** [reply]
@GytisZ Thanks for reaching out! At the moment, Copilot Chat in your IDE will only use your local files - unless you're using the [`@workspace` feature in VS Code](https://code.visualstudio.com/blogs/2023/11/13/vscode-copilot-smarter), in which case the situation is a bit more complex and GitHub-indexed repos can be used.

We're definitely aware of the need to bring the power of Copilot Enterprise to your editor, and this is something we're actively thinking about and working on - watch this space!

What kind of functionality would you want to see in your IDE?

**Comment by vannela** [comment]
### **Content exclusion:**

Does Copilot Enterprise support Azure DevOps or Azure DevOps Server (TFS) repositories for using "Content Exclusion" feature? 
If so, please share guidance on how to configure it?

**Reply by timrogers** [reply]
👋🏻 Copilot Enterprise only works with content stored in GitHub. It does not support data stored in Azure DevOps or Azure DevOps Server. 

Can you share any more details on how you would want Azure DevOps integration to work?

**Reply by vannela** [reply]
It seems the response is counter to some of the responses in question 1 above. that through IDE (like VS Code or VS) integration of copilot enterprise works with any type of repository.
 
If above statement is true, my question is about ways to exclude any of the content from repositories like Azure DevOps or TFS?

**Reply by timrogers** [reply]
Ah, I think I understand your question now - apologies for the misunderstanding!

You can configure Content Exclusions to exclude Azure DevOps or Azure DevOps Server repos based on the repository URL: https://docs.github.com/en/copilot/managing-github-copilot-in-your-organization/configuring-content-exclusions-for-github-copilot#what-can-you-exclude

This only applies to code completions, and not to other IDE features like Copilot Chat.

**Comment by jcleigh** [comment]
Hello, I have two questions:

1. Is upgrading from Copilot Business to Enterprise an org-wide change or can it be done per-seat?
2. Can we downgrade back to Business if we do not find value in Enterprise?

Thanks in advance!

**Reply by jcleigh** [reply]
GitHub support responded to my inquiry so I'll share here as well:

> Can we upgrade on a per-seat basis or is it org-wide?

The entire organization would need to be upgraded to the Enterprise plan, but not every member needs to have Copilot Enterprise or Business enabled. You can choose which members you'd specifically like to have access.
It is not possible to have some users on Copilot Business and some on Copilot Enterprise, they must all be on one or the other.

> If we do not find value in Copilot Enterprise, are we able to downgrade back to Copilot Business?

Yes, though currently there's not a way to do this specific downgrade in the app, if you reach out to Support we'll be glad to help with the switch back to Copilot Business.

**Comment by KKrzeszowska** [comment]
You provide support and defence of third-party claims in case the use of a Suggestion of GitHub Copilot violates the third-party rights. Does it apply to the case when we purchase GitHub license for our devs and write code as a contractor for our clients? Does it cover such cases in which the owner of the code in the end is a third party (our client)?

**Reply by timrogers** [reply]
@KKrzeszowska Thanks for reaching out! I'm working to get an answer to this for you internally.

**Comment by Beanmasters** [comment]
Can you expand on what "Answers based on your organization’s knowledge base" included in the Enterprise subscription means are and what that works with?  For example, does this mean Copilot can incorporate the information from a private Atlassian Confluence cloud instance in its answers?

https://github.com/features/copilot?ef_id=_k_41f20b113d2311eaf5286ee096924aaf_k_&OCID=AIDcmmc3fhtaow_SEM__k_41f20b113d2311eaf5286ee096924aaf_k_&msclkid=41f20b113d2311eaf5286ee096924aaf

**Reply by timrogers** [reply]
👋🏻 Hi @Beanmasters! Knowledge bases allow you to [ask questions](https://docs.github.com/en/enterprise-cloud@latest/copilot/github-copilot-enterprise/copilot-chat-in-github/using-github-copilot-chat-in-githubcom#asking-a-question-about-a-knowledge-base) about Markdown documentation stored in one or more GitHub repositories. [Organization owners can set up knowledge bases](https://docs.github.com/en/enterprise-cloud@latest/copilot/github-copilot-enterprise/copilot-chat-in-github/managing-copilot-knowledge-bases) and then they're shared with the whole organization.

At this time, we don't support external sources of knowledge like Confluence, but this is something we're considering for the future - plus cases like this will be enabled by our [upcoming Copilot Partner Program](https://github.blog/2023-11-08-universe-2023-copilot-transforms-github-into-the-ai-powered-developer-platform/#fostering-an-ecosystem-built-on-copilot-with-the-github-copilot-partner-program).

**Reply by MCraenen** [reply]
Hi @timrogers! On that topic, are there plans to support adding repositories hosted outside of Github (e.g. in Gitlab) as parts of Copilot knowledge bases?

**Reply by timrogers** [reply]
@MCraenen We don't have any plans to do that at the moment. Is it just GitLab you're thinking of, or other sources of knowledge as well?

**Reply by MCraenen** [reply]
Thanks for the swift reply @timrogers! It's just Gitlab I am thinking of for my use case. We self-host our gitlab instances but use Copilot often in our IDEs. Knowledge bases seem like they'd be a super useful extension to our copilot use cases especially now with VSCode support, but if all the code has to be hosted on github we unfortunately can't use it.

**Comment by simensol** [comment]
I’m using Visual Studio Code with GitHub Copilot, and I see two chat participants: `@GitHub` and `@workspace`. My workspace is my repository, which is already indexed. Could you explain the differences between these two participants? Specifically, what are the benefits of using `@GitHub` compared to `@workspace` in this scenario?

**Comment by adriaanse-deltares** [comment]
My organization has upgraded a group of developers from the business edition to copilot enterprise after reading that this product is generally available. However after setting it all up we discover that the main functionality of creating "custom models" is missing from the admin UI and after placing a support call we are told: 

> The custom models for GitHub Copilot are now available in public preview. If you haven't done so already, please join the waitlist to gain access to the Custom Models. Also from the screenshot shared, you need to enable "Opt in for access to Copilot preview features" which would allow your organization preview new Copilot in github.com feature.
>  
> We are unable to provide an estimated timeline as the criteria for selection are determined by the engineering team and may evolve as the beta progresses. GitHub Support does not have the ability to expedite or influence the selection process, If your organization is selected, you’ll be notified directly, so please make sure your primary email address is up-to-date.
> 

It is already months ago that we saw this announcement, which according to support simply is not true, and we have been wasting our time.
A great way to lose your "early adapter" clients I would say, I feel kind of ripped off !  Is this because we are in Europe ?

https://github.blog/news-insights/product-news/github-copilot-enterprise-is-now-generally-available/

**Reply by MarcLupiane** [reply]
We have the same problem. We bought Github Enterprise for the solve reason to use Copilot with "Custom Models". On [Github Pricing](https://github.com/pricing) and [Copilot Plans](https://github.com/features/copilot/plans) it clearly stated, that you can use the Copilot Custom Models feature with the Enterprise version. We set up everything according to the requirements, but the feature does not seem to be enabled.

Now a few days later, the "Custom Model" feature disappeared everywhere. I feel ripped off as well.

**Reply by adriaanse-deltares** [reply]
It is still listed here: 
https://docs.github.com/en/copilot/about-github-copilot/subscription-plans-for-github-copilot
https://docs.github.com/en/enterprise-cloud@latest/copilot/customizing-copilot/creating-a-custom-model-for-github-copilot

But thanks for the heads up, I'll keep an eye on it to see if those pages will be removed also, we are now looking into Codeium as alternative....

**Reply by MarcLupiane** [reply]
Ok, so now they added the "**limited** public preview"....
